UnityProject/Assets/Scripts/Objects/Atmospherics/AirController.cs

```diff
@@ -0,0 +1,357 @@
+				"Disable if the ACU is not in the room it controls. Important to disable for Atmospherics reservoir ACUs.")]
+		private bool acuSamplesAir = true;
+
+		private MetaDataNode facingMetaNode;
+		private AccessRestrictions accessRestrictions;
+		private SpriteHandler spriteHandler;
+
+		/// <summary>Invoked when the air controller's state changes.</summary>
+		public Action OnStateChanged;
+
+		public readonly HashSet<IAcuControllable> ConnectedDevices = new HashSet<IAcuControllable>();
+
+		/// <summary>
+		/// The mode this controller and the devices it controls should operate with,
+		/// given other conditions, like being powered, are met.
+		/// </summary>
+		public AcuMode DesiredMode { get; private set; }
+
+		/// <summary>
+		/// We piggyback off a <c>GasMix</c> to represent the average <c>GasMix</c> over all devices, including the controller.
+		/// <para>For data representation only.</para>
+		/// </summary>
+		public GasMix AverageGasMix { get; private set; }
+
+		public AcuThresholds Thresholds { get; private set; }
+
+		public bool IsWriteable => IsPowered && IsLocked == false;
+
+		#region Lifecycle
+
+		private bool IsReady => facingMetaNode != null;
+
+		private void Awake()
+		{
+			accessRestrictions = GetComponent<AccessRestrictions>();
+			spriteHandler = GetComponentInChildren<SpriteHandler>();
+
+			Thresholds = initialAcuThresholds.Clone();
+			GasLevelStatus = new AcuStatus[Gas.Gases.Count];
+			DesiredMode = initialOperatingMode;
+		}
+
+		private void OnDisable()
+		{
+			UpdateManager.Remove(CallbackType.PERIODIC_UPDATE, PeriodicUpdate);
+		}
+
+		public void OnSpawnServer(SpawnInfo info)
+		{
+			var registerTile = gameObject.RegisterTile();
+			var localPos = registerTile.LocalPositionServer + GetComponent<WallmountBehavior>().CalculateFacing();
+			facingMetaNode = registerTile.Matrix.MatrixInfo.MetaDataLayer.Get(localPos.CutToInt());
+
+			if (IsPowered)
+			{
+				UpdateManager.Add(PeriodicUpdate, 3);
+				PeriodicUpdate();
+			}
+		}
+
+		#endregion
+
+		private void PeriodicUpdate()
+		{
+			AverageGasMix = GetAverageGasMix();
+			UpdateStatusProperties();
+			spriteHandler.ChangeSprite((int)OverallStatus);
+
+			// Cycle will vacuum air, and then refill.
+			if (DesiredMode == AcuMode.Cycle && AverageGasMix.Pressure < AtmosConstants.ONE_ATMOSPHERE / 20)
+			{
+				SetOperatingMode(AcuMode.Refill);
+			}
+			// Slightly different to /tg/: refill will now return to the initial operating mode once one atmosphere is reached.
+			else if (DesiredMode == AcuMode.Refill && AverageGasMix.Pressure > AtmosConstants.ONE_ATMOSPHERE)
+			{
+				SetOperatingMode(initialOperatingMode);
+			}
+		}
+
+		private GasMix GetAverageGasMix()
+		{
+			int sampleSourceCount = ConnectedDevices.Count + (acuSamplesAir ? 1 : 0);
+
+			// Sample the gas at the ACU (if enabled).
+			GasMix avgMix = GasMix.NewGasMix(acuSamplesAir ? facingMetaNode.GasMix : GasMixes.BaseEmptyMix);
```

Comment:
Might be a GC issue to create a new gas mix every 3 seconds for all AirAlarms, might wanna see if we can create one at spawn then clear and reuse.